lib/src/importer/node_package.dart

```diff
@@ -0,0 +1,362 @@
+/// canonicalizing paths not defined in package.json.
+final _filesystemImporter = FilesystemImporter.cwd;
+
+/// An [Importer] that resolves `pkg:` URLs using the Node resolution algorithm.
+class NodePackageImporterInternal extends Importer {
+  final Uri entryPointURL;
+
+  /// Creates a Node Package Importer with the associated entry point url
+  NodePackageImporterInternal(this.entryPointURL);
+
+  @override
+  bool isNonCanonicalScheme(String scheme) {
+    return scheme == 'pkg';
+  }
+
+  @override
+  Uri? canonicalize(Uri url) {
+    if (url.scheme == 'file') return _filesystemImporter.canonicalize(url);
+    if (url.scheme != 'pkg') return null;
+
+    if (url.hasAuthority) {
+      throw "pkg: URL $url must not have a host, port, username or password.";
+    }
+    if (p.isAbsolute(url.path)) {
+      throw "pkg: URL $url must not be an absolute path.";
+    }
+    if (url.path.isEmpty) {
+      throw "pkg: URL $url must not have an empty path.";
+    }
+    if (url.hasQuery || url.hasFragment) {
+      throw "pkg: URL $url must not have a query or fragment.";
+    }
+
+    var baseURL = containingUrl?.scheme == 'file'
+        ? Uri.parse(containingUrl!.toFilePath())
+        : entryPointURL;
+
+    var (packageName, subpath) = _packageNameAndSubpath(url.path);
+    var packageRoot = _resolvePackageRoot(packageName, baseURL);
+    if (packageRoot == null) {
+      return null;
+    }
+
+    // Attempt to resolve using conditional exports
+    var jsonPath = p.join(packageRoot.toFilePath(), 'package.json');
+    var jsonFile = Uri.file(jsonPath).toFilePath();
+
+    var jsonString = readFile(jsonFile);
+    var packageManifest = jsonDecode(jsonString) as Map<String, dynamic>;
+
+    var resolved = _resolvePackageExports(
+        packageRoot, subpath, packageManifest, packageName);
+
+    if (resolved != null &&
+        resolved.scheme == 'file' &&
+        ['.scss', '.sass', '.css'].contains(p.extension(resolved.path))) {
+      return resolved;
+    } else if (resolved != null) {
+      throw "The export for '${subpath == '' ? "root" : subpath}' in "
+          "'$packageName' is not a valid Sass file.";
+    }
+    // If no subpath, attempt to resolve `sass` or `style` key in package.json,
+    // then `index` file at package root, resolved for file extensions and
+    // partials.
+    if (subpath == '') {
+      return _resolvePackageRootValues(
+          packageRoot.toFilePath(), packageManifest);
+    }
+
+    // If there is a subpath, attempt to resolve the path relative to the
+    // package root, and resolve for file extensions and partials.
+    var relativeSubpath = "${packageRoot.toFilePath()}${p.separator}$subpath";
+    return _filesystemImporter.canonicalize(Uri.file(relativeSubpath));
+  }
+
+  @override
+  ImporterResult? load(Uri url) => _filesystemImporter.load(url);
+
+  /// Takes a string, `path`, and returns a tuple with the package name and the
```

Comment:
I looked a bit more into it, and there are some parts where `/` is required, and some where either would be allowed. 

- Export keys _must_ start with a `./`
- Package name and subpath _must_ be split on `/`.  
- The package name must _not_ contain a `/`.
- The export key must _exactly_ match the subpath.

The one exception that I'm seeing based on the Node source code, it's possible that a path like `package/src\index.scss` could match to an export of `./src\index.scss` in `package`, but since the pkg URL can't have backslashes, we don't need to cover that edge case.

I switched to using path.posix, and threw a couple more errors in ed585906ec0ee6b57518d83d03c5e872e6958402